src/amber/controller/redirect.cr

```diff
@@ -1,61 +1,68 @@
 module Amber::Controller
-  # This class writes the redirect URL to the response headers and parses params.
-  class LocationRedirect
+  module RedirectMethods
+    def redirect_to(location : String, **args)
+      Redirector.new(location, **args).redirect(self)
+    end
+
+    def redirect_to(action : Symbol, **args)
+      Redirector.from_controller_action(controller_name, action, **args).redirect(self)
+    end
+
+    def redirect_to(controller : Symbol, action : Symbol, **args)
+      Redirector.from_controller_action(controller.to_s, action, **args).redirect(self)
+    end
+
+    def redirect_back(**args)
+      Redirector.new(request.headers["Referer"].to_s, **args).redirect(self)
+    end
+  end
+
+  class Redirector
     getter location, status, params, flash
 
     @location : String
     @status : Int32 = 302
     @params : Hash(String, String)? = nil
     @flash : Hash(String, String)? = nil
 
-    def initialize(@location, @status = 302, @params = nil, @flash = nil)
-      raise_redirect_error(location) if location.empty?
+    def self.from_controller_action(controller : String, action : Symbol, **options)
+      router = Amber::Router::Router.instance
+      route = router.match_by_controller_action(controller, action)
+      if params = options[:params]?
+        location, params = route.not_nil!.parse_params(params)
+        status = options[:status]? || 302
+        new(location, status: status, params: params, flash: options[:flash]?)
+      else
+        raise Exceptions::Controller::Redirect.new("#{controller}##{action} not found!")
+      end
     end
 
-    def redirect(for context)
-      response = context.response
-      context.flash.merge!(flash.not_nil!) if !flash.nil?
-      raise_redirect_error(location) unless location.url? || location.chars.first == '/'
-      set_location(response, location, status, params)
+    def initialize(@location, @status = 302, @params = nil, @flash = nil)
+      raise_redirect_error(location) if location.empty?
     end
 
-    private def set_location(response, location, status = 302, params : _ = nil)
+    def redirect(controller)
+      set_flash(controller)
       url_path = encode_query_string(location, params)
-      response.headers.add "Location", url_path
-      response.status_code = status
+      controller.response.headers.add "Location", url_path
+      controller.halt!(status, "Redirecting to #{url_path}")
     end
 
-    def encode_query_string(location, params)
-      return location + "?" + HTTP::Params.encode(params).to_s if params
+    private def encode_query_string(location, params)
+      if !params.nil? && !params.empty?
```

Comment:
Unless for some reason raises a compile error on `params.empty?` **Undefined method 'empty?' for Nil**, forcing me to rewrite it aas `params.to_s.empty?`.